test/parallel_api/algorithm/alg.sorting/sort_by_key_common.h

```diff
@@ -0,0 +1,250 @@
+#   include "support/sycl_alloc_utils.h"
+#endif
+
+#include <vector>
+#include <algorithm>
+#include <type_traits>
+
+struct StableSort{};
+struct UnstableSort{};
+
+struct AscendingSort{};
+struct DiscendingSort{};
+
+template<typename TestType, typename StabilityTag, typename DirectionTag>
+struct KernelName{};
+
+template<typename _KeyIt, typename _ValIt, typename _Size, typename DirectionTag>
+void
+generate_data(_KeyIt keys_begin, _ValIt vals_begin, _Size n, DirectionTag direction_tag)
+{
+    if constexpr (std::is_same_v<DirectionTag, AscendingSort>)
+    {
+        // Generated example for n = 10:
+        // Keys:    1 2 1 2 1 2 1 2 1 2
+        // Values:  0 1 2 3 4 5 6 7 8 9
+        //
+        // Sorted example for n = 10, stable sort:
+        // Keys:    1 1 1 1 1 2 2 2 2 2
+        // Values:  0 2 4 6 8 1 3 5 7 9
+        auto counting_begin = oneapi::dpl::counting_iterator<int>{0};
+        std::transform(counting_begin, counting_begin + n, keys_begin,
+                    [](int i) { return i % 2 + 1; });
+        std::copy(counting_begin, counting_begin + n, vals_begin);
+    }
+    else
+    {
+        // Generated example for n = 10:
+        // Keys:    0 10 20 30 40 50 60 70 80 90
+        // Values:  9  8  7  6  5  4  3  2  1  0
+        //
+        // Sorted example for n = 10, stable sort:
+        // Keys:    90 80 70 60 50 40 30 20 10  0
+        // Values:   0  1  2  3  4  5  6  7  8  9
+        for (int i = 0; i < n; i++)
+        {
+            keys_begin[i] = i * 10;
+            vals_begin[i] = n - i - 1;
+        }
+    }
+}
+
+template<typename _Policy, typename _KeyIt, typename _ValIt, typename _Size, typename StabilityTag, typename DirectionTag>
+void
+call_sort(_Policy&& policy, _KeyIt keys_begin, _ValIt vals_begin, _Size n, StabilityTag, DirectionTag)
+{
+    if constexpr (std::is_same_v<DirectionTag, AscendingSort>)
+    {
+        // Do not pass the comparator to check the API with a default comparator; radix sort may be used
+        if constexpr (std::is_same_v<StabilityTag, StableSort>)
+            oneapi::dpl::stable_sort_by_key(policy, keys_begin, keys_begin + n, vals_begin);
+        else
+            oneapi::dpl::sort_by_key(policy, keys_begin, keys_begin + n, vals_begin);
+    }
+    else
+    {
+        // Pass a custom comparator to check the corresponding implementation; merge-sort may be used
+        auto greater = [](const auto& lhs, const auto& rhs) { return lhs > rhs; };
+        if constexpr (std::is_same_v<StabilityTag, StableSort>)
+            oneapi::dpl::stable_sort_by_key(policy, keys_begin, keys_begin + n, vals_begin, greater);
+        else
+            oneapi::dpl::sort_by_key(policy, keys_begin, keys_begin + n, vals_begin, greater);
+    }
+}
+
+template<typename _KeysIt, typename _ValsIt, typename _Size, typename StabilityTag, typename DirectionTag>
+void
+check_sort(const _KeysIt& keys_begin, const _ValsIt& vals_begin, _Size n, StabilityTag, DirectionTag)
```

Comment:
A checker depends on a data generator, right?
Why whether do we use `std::is_sorted`? or/and std::sort with tuple type and check via EXPECT_EQ_N ?

Reply:
I've made a more genetic checkers and generator, so they are now decoupled. `is_sorted` is not used now either.
Has it resolved your question? Frankly, I do not understand all of its details.

Reply:
> The last sentence tells: if a data generator is changed, the checker should be also changed (checker depends on data generation way). It is a downside of the test design.

By the way, the checker and generator do not depend on each other now. The current issue is that they are embedded - other variants cannot be provided unless the code structure is changed, but it is not necessary right now.